src/boot/efi/xbootldr.c

```diff
@@ -243,9 +254,7 @@ static EFI_STATUS find_device(EFI_HANDLE *device, EFI_DEVICE_PATH **ret_device_p
                 }
 
                 /* Patch in the data we found */
-                EFI_DEVICE_PATH *xboot_path = path_dup(partition_path);
-                memcpy((uint8_t *) xboot_path + ((uint8_t *) part_node - (uint8_t *) partition_path), &hd, sizeof(hd));
-                *ret_device_path = xboot_path;
+                *ret_device_path = path_replace(partition_path, part_node, (EFI_DEVICE_PATH *) &hd);
```

Comment:
hmm, using `&hd->header` as third arg should work too, no? and not require an ugly cast...

Reply:
Should work, but that would make it look like it only operates on the header when it really works on the whole struct. That looks just wrong to me.

Reply:
but now it looks like you cast the pointer to something entirely unrelated...